### PR TITLE
Support for pretrained word embeddings

### PR DESCRIPTION
**Pull request by may-**
- typo fix
- bug fix: scripts/average_checkpoints.py
- #112 load pretrained embeddings from file

--- 
- pretrained embeddings file has to have the format:
   Example:
   ```
   2 5
   the -0.0230 -0.0264  0.0287  0.0171  0.1403
   at -0.0395 -0.1286  0.0275  0.0254 -0.0932
   ``` 
    - the first line should have two integers: vocabulary size and dimension
    - each line should contain word and embedding weights separated by spaces.
    - dimension has to match with config (`config["model"]["encoder"]["embeddings"]["embedding_dim"]`)
- I tested `configs/iwslt_deen_bahdanau.yaml` with pretrained embeddings created as follows:
  - I used [fasttext](https://github.com/facebookresearch/fastText/). clone the repo and install it. (The file format described above is more or less conventional across frameworks, though. You can use i.e. [gensim](https://github.com/RaRe-Technologies/gensim) instead.)
  - download pretrained models and reduce the dimension
     ```
     $ cd fastText
     $ ./download_model.py en         # will download cc.en.300.bin
     $ ./download_model.py de         # will download cc.de.300.bin
     $ ./reduce_model.py cc.en.300.bin 256         # will create cc.de.256.bin
     $ ./reduce_model.py cc.de.300.bin 256         # will create cc.de.256.bin
     ```
  - apply to the train data and add headers
     ```
     $ cd test/data/iwslt14_deen/      # data dir
     $ fasttext print-word-vectors cc.en.256.bin < train.en > train.en.embed
     $ fasttext print-word-vectors cc.de.256.bin < train.de > train.de.embed
     $ wc -l train.en.embed
     3191644 train.en.embed
     $ wc -l train.de.embed
     3022820 train.de.embed
     $ echo "3191644 256" | cat - train.en.embed > train.en.embed.txt
     $ echo "3022820 256" | cat - train.de.embed > train.de.embed.txt
     ```
     - I haven't checked any tokenization issue. ideally, you should use the same tokenization method as the pretrained models.
     - for joined vocabulary, you can just concatenate these txt files (only one header line, though!)
  - then put this file path to the `config.yaml`
     ```yaml
     model:
        encoder:
            embeddings:
                embedding_dim: 256
                load_pretrained: "test/data/iwslt14_deen/train.de.embed.txt"
        decoder:
            embeddings:
                embedding_dim: 256
                load_pretrained: "test/data/iwslt14_deen/train.en.embed.txt"
     ```
- Note that the weights will be overwitten if you specify `config["training"]["load_model"]` and the model has embeddings layers.

- BLEU `configs/iwslt_deen_bahdanau.yaml` (beam size = 10)  
   ||dev|test|
   |:-|:-|:-|
   |w/o pretrained embeddings|30.13|29.40|
   |with pretrained embeddings|29.87|29.45|

   
learning_curves

### CODE REVIEW

File: configs/small.yaml [hunk 1]
```diff
@@ -98,6 +99,7 @@ model:                              # specify your model architecture here
             embedding_dim: 16
             scale: False
             freeze: False           # if True, embeddings are not updated during training
```
Comment (juliakreutzer) [inline configs/small.yaml:101]: Have you tried loading the fasttext embeddings and then freezing them for IWSLT?

File: joeynmt/embeddings.py [hunk 2]
```diff
@@ -53,3 +61,39 @@ def forward(self, x: Tensor) -> Tensor:
     def __repr__(self):
         return "%s(embedding_dim=%d, vocab_size=%d)" % (
             self.__class__.__name__, self.embedding_dim, self.vocab_size)
+
+    #from fairseq
+    def load_from_file(self, embed_path: str, vocab: Vocabulary):
+        """Load pretrained embedding weights from text file.
+
+        - first line will be ignored (vocabulary size and dimension)
```
Comment (juliakreutzer) [inline joeynmt/embeddings.py:69]: Maybe clarify: First line is expected to contain vocabulary size and dimension. The dimension has to match the model's specified embedding size, the vocabulary size is not used.

Comment (juliakreutzer) [reply joeynmt/embeddings.py:69]: It would also be great to add a line about what happens with vocabulary items that are not part of the pretrained vocabulary.

File: joeynmt/embeddings.py [hunk 3]
```diff
@@ -53,3 +61,39 @@ def forward(self, x: Tensor) -> Tensor:
...
+            the -0.0230 -0.0264  0.0287  0.0171  0.1403
+            at -0.0395 -0.1286  0.0275  0.0254 -0.0932
+
+        :param embed_path: embedding weights text file
+        :param vocab: Vocabulary object
+        """
+        embed_dict = {}
+        # parse file
+        with io.open(embed_path, 'r', encoding='utf-8',
+                     errors='ignore') as f_embed:
```
Comment (juliakreutzer) [inline joeynmt/embeddings.py:85]: I'm wondering if we could log the number of tokens of the model vocab that were matched, e.g. "found 899 of 1000 (89.9%) tokens of the vocabulary in the pre-trained embeddings", so that if tokenization is different, e.g. contains BPEs, the user might notice a very low number of matches there.

File: joeynmt/model.py [hunk 4]
```diff
@@ -209,6 +212,7 @@ def build_model(cfg: dict = None,
     :param trg_vocab: target vocabulary
     :return: built and initialized model
     """
+    logger.info("building an encoder-decoder model...")
```
Comment (juliakreutzer) [inline joeynmt/model.py:215]: Tiny detail: Start logging messages with uppercase letters.

**Review by juliakreutzer — CHANGES_REQUESTED** [review]

**Comment by juliakreutzer** [comment]
That's a wonderful addition, thank you @may- !

File: configs/small.yaml (hunk 1 above)
Comment (may-) [reply configs/small.yaml:101]: No, I haven't tried freezing. I've specified neither "scale" nor "freeze" option in config, that is, default values (scale: False, freeze: False) are used, so I just initialized the WE by the pre-trained weights, nothing more. Do you simply want to see the results with freezing? Or have I made any mistake, i.e. I should always freeze WE when I initialize it by pretrained weights??

Comment (juliakreutzer) [reply configs/small.yaml:101]: I was just curious about results :) no mistakes.

Comment (juliakreutzer) [reply configs/small.yaml:101]: I was also thinking that renormalization with the randomly initialised OOV words would be good so that all embeddings are on the same scale, but we can keep for later. It might be worth making it a default if it turns out to be bénéficient.

File: joeynmt/embeddings.py (hunk 3 above)
Comment (may-) [reply joeynmt/embeddings.py:85]: That log message might be a bit ambiguous. Let's say `intersection` = num tokens found both in the pretrained file and joeynmt's vocab. 
- len(intersection) / len(vocab size in pretrained file)
- len(intersection) / len(joeynmt's vocab.stoi)

I guess you are suggesting to log the latter, aren't you?

File: configs/small.yaml (hunk 1 above)
Comment (may-) [reply configs/small.yaml:102]: oh I see. Maybe we can at least raise attention in doc: loading pretrained weights may cause scaling issue, users should be careful on choosing WE initialization.

I'll run extra experiments with freezing and report results here, when we have free space in our GPUs. (Recently, our gpus are occupied...)

Comment (juliakreutzer) [reply configs/small.yaml:102]: yes that sounds like a very good solution. Don't worry about the experiments. What do you think about making a small tutorial of it? With exactly the experiments you ran so far -- I think it's worth having an extra page in the documentation for it (like the toy task), since it's a common use case.

File: joeynmt/embeddings.py (hunk 3 above)
Comment (juliakreutzer) [reply joeynmt/embeddings.py:95]: Yes, exactly. Perhaps a better formulation: "Found 899 of 1000 (89.9%) tokens of the NMT vocabulary in the pre-trained embeddings, the remaining ones are initialized randomly."

**Review by juliakreutzer — APPROVED** [review]
Looks very good, thank you for the adjustments!

**Comment by juliakreutzer** [comment]
Just to reiterate on the comment I had for a conversation that I marked now as resolved (not sure if you still got the message): What do you think about making a small tutorial of it? With exactly the experiments you ran so far -- I think it's worth having an extra page in the documentation for it (like the toy task), since it's a common use case. I know this takes extra time, but just giving the command line instructions and showing where to download the embeddings would be very helpful I think.

**Comment by may-** [comment]
@juliakreutzer ah, ok. I'll write a small tutorial for it. (maybe in March or April... )

**Comment by may-** [comment]
just a small note: looking at the translation by a model on pretrained embeddings with `freeze=True` option, the model is struggling to learn `<unk>`, although the model reached almost the same level of corpus bleu score (ca. 30 bleu).  I don't know how sacrebleu handles hypotheses with `<unk>`. Maybe you need to de-normalize `<unk>` output before bleu calculation. or, pretrained "subword" embeddings (instead of "word" embeddings) would solve this problem. I'm not sure.

**Comment by juliakreutzer** [comment]
Interesting! What do you mean by "struggling to learn <unk>"? Does it generate too many unks? Removing unks before BLEU would probably be a good hack.
